File: zig/src/data_structures/doubly_linked_list.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

/// Doubly linked list with bidirectional traversal.
/// Time: O(1) for prepend/append/removeFirst/removeLast
/// Space: O(n)
pub fn DoublyLinkedList(comptime T: type) type {
    return struct {
        const Self = @This();
        const Node = struct {
            value: T,
            prev: ?*Node,
            next: ?*Node,
        };

        allocator: Allocator,
        head: ?*Node,
        tail: ?*Node,
        count: usize,

        pub fn init(allocator: Allocator) Self {
            return .{
                .allocator = allocator,
                .head = null,
                .tail = null,
                .count = 0,
            };
        }

        pub fn deinit(self: *Self) void {
            var current = self.head;
            while (current) |node| {
                const next = node.next;
                self.allocator.destroy(node);
                current = next;
            }
        }

        pub fn prepend(self: *Self, value: T) !void {
            const node = try self.allocator.create(Node);
            node.* = .{ .value = value, .prev = null, .next = self.head };

            if (self.head) |head| {
                head.prev = node;
            } else {
                self.tail = node;
            }
            self.head = node;
            self.count += 1;
        }

        pub fn append(self: *Self, value: T) !void {
            const node = try self.allocator.create(Node);
            node.* = .{ .value = value, .prev = self.tail, .next = null };

            if (self.tail) |tail| {
                tail.next = node;
            } else {
                self.head = node;
            }
            self.tail = node;
            self.count += 1;
        }

        pub fn removeFirst(self: *Self) ?T {
            const head = self.head orelse return null;
            const value = head.value;

            self.head = head.next;
            if (self.head) |new_head| {
                new_head.prev = null;
            } else {
                self.tail = null;
            }

            self.allocator.destroy(head);
            self.count -= 1;
            return value;
        }

        pub fn removeLast(self: *Self) ?T {
            const tail = self.tail orelse return null;
            const value = tail.value;

            self.tail = tail.prev;
            if (self.tail) |new_tail| {
                new_tail.next = null;
            } else {
                self.head = null;
            }

            self.allocator.destroy(tail);
            self.count -= 1;
            return value;
        }

        pub fn first(self: *const Self) ?T {
            return if (self.head) |head| head.value else null;
        }

        pub fn last(self: *const Self) ?T {
            return if (self.tail) |tail| tail.value else null;
        }

        pub fn get(self: *const Self, index: usize) ?T {
            if (index >= self.count) return null;

            var current: ?*Node = undefined;
            if (index < self.count / 2) {
                current = self.head;
                var i: usize = 0;
                while (i < index) : (i += 1) {
                    current = current.?.next;
                }
            } else {
                current = self.tail;
                var i: usize = self.count - 1;
                while (i > index) : (i -= 1) {
                    current = current.?.prev;
                }
            }
            return current.?.value;
        }

        pub fn isEmpty(self: *const Self) bool {
            return self.count == 0;
        }

        pub fn size(self: *const Self) usize {
            return self.count;
        }
    };
}

test "DoublyLinkedList operations" {
    const allocator = std.testing.allocator;
    var list = DoublyLinkedList(i32).init(allocator);
    defer list.deinit();

    try list.prepend(2);
    try list.prepend(1);
    try list.append(3);

    try std.testing.expectEqual(@as(?i32, 1), list.first());
    try std.testing.expectEqual(@as(?i32, 3), list.last());
    try std.testing.expectEqual(@as(usize, 3), list.size());
}

File: rust/src/data_structures/doubly_linked_list.rs
use std::cell::RefCell;
use std::rc::Rc;

/// A doubly linked list implementation.
/// Time: O(1) for push/pop front/back, O(n) for search
/// Space: O(n)

type Link<T> = Option<Rc<RefCell<Node<T>>>>;

#[derive(Debug)]
struct Node<T> {
    value: T,
    next: Link<T>,
    prev: Link<T>,
}

#[derive(Debug)]
pub struct DoublyLinkedList<T> {
    head: Link<T>,
    tail: Link<T>,
    len: usize,
}

impl<T> DoublyLinkedList<T> {
    pub fn new() -> Self {
        DoublyLinkedList {
            head: None,
            tail: None,
            len: 0,
        }
    }

    pub fn push_front(&mut self, value: T) {
        let new_node = Rc::new(RefCell::new(Node {
            value,
            next: self.head.clone(),
            prev: None,
        }));

        match self.head.take() {
            Some(old_head) => {
                old_head.borrow_mut().prev = Some(new_node.clone());
            }
            None => {
                self.tail = Some(new_node.clone());
            }
        }

        self.head = Some(new_node);
        self.len += 1;
    }

    pub fn push_back(&mut self, value: T) {
        let new_node = Rc::new(RefCell::new(Node {
            value,
            next: None,
            prev: self.tail.clone(),
        }));

        match self.tail.take() {
            Some(old_tail) => {
                old_tail.borrow_mut().next = Some(new_node.clone());
            }
            None => {
                self.head = Some(new_node.clone());
            }
        }

        self.tail = Some(new_node);
        self.len += 1;
    }

    pub fn pop_front(&mut self) -> Option<T> {
        self.head.take().map(|old_head| {
            match old_head.borrow_mut().next.take() {
                Some(new_head) => {
                    new_head.borrow_mut().prev = None;
                    self.head = Some(new_head);
                }
                None => {
                    self.tail = None;
                }
            }
            self.len -= 1;
            Rc::try_unwrap(old_head).ok().unwrap().into_inner().value
        })
    }

    pub fn pop_back(&mut self) -> Option<T> {
        self.tail.take().map(|old_tail| {
            match old_tail.borrow_mut().prev.take() {
                Some(new_tail) => {
                    new_tail.borrow_mut().next = None;
                    self.tail = Some(new_tail);
                }
                None => {
                    self.head = None;
                }
            }
            self.len -= 1;
            Rc::try_unwrap(old_tail).ok().unwrap().into_inner().value
        })
    }

    pub fn peek_front(&self) -> Option<T>
    where
        T: Clone,
    {
        self.head.as_ref().map(|node| node.borrow().value.clone())
    }

    pub fn peek_back(&self) -> Option<T>
    where
        T: Clone,
    {
        self.tail.as_ref().map(|node| node.borrow().value.clone())
    }

    pub fn len(&self) -> usize {
        self.len
    }

    pub fn is_empty(&self) -> bool {
        self.len == 0
    }

    pub fn clear(&mut self) {
        while self.pop_front().is_some() {}
    }
}

impl<T> Default for DoublyLinkedList<T> {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_push_pop_front() {
        let mut list = DoublyLinkedList::new();
        list.push_front(1);
        list.push_front(2);
        list.push_front(3);

        assert_eq!(list.pop_front(), Some(3));
        assert_eq!(list.pop_front(), Some(2));
        assert_eq!(list.pop_front(), Some(1));
        assert_eq!(list.pop_front(), None);
    }

    #[test]
    fn test_push_pop_back() {
        let mut list = DoublyLinkedList::new();
        list.push_back(1);
        list.push_back(2);
        list.push_back(3);

        assert_eq!(list.pop_back(), Some(3));
        assert_eq!(list.pop_back(), Some(2));
        assert_eq!(list.pop_back(), Some(1));
        assert_eq!(list.pop_back(), None);
    }

    #[test]
    fn test_mixed_operations() {
        let mut list = DoublyLinkedList::new();
        list.push_back(1);
        list.push_back(2);
        list.push_front(0);

        assert_eq!(list.pop_front(), Some(0));
        assert_eq!(list.pop_back(), Some(2));
        assert_eq!(list.pop_front(), Some(1));
    }

    #[test]
    fn test_peek() {
        let mut list = DoublyLinkedList::new();
        list.push_back(1);
        list.push_back(2);

        assert_eq!(list.peek_front(), Some(1));
        assert_eq!(list.peek_back(), Some(2));
    }

    #[test]
    fn test_len() {
        let mut list = DoublyLinkedList::new();
        assert!(list.is_empty());

        list.push_back(1);
        list.push_back(2);
        assert_eq!(list.len(), 2);
    }
}
